src/Symfony/Component/Security/Http/Firewall/ContextListener.php

```diff
@@ -292,9 +292,13 @@ private static function hasUserChanged(UserInterface $originalUser, TokenInterfa
         }
 
         if ($originalUser instanceof PasswordAuthenticatedUserInterface || $refreshedUser instanceof PasswordAuthenticatedUserInterface) {
-            if (!$originalUser instanceof PasswordAuthenticatedUserInterface
-                || !$refreshedUser instanceof PasswordAuthenticatedUserInterface
-                || $refreshedUser->getPassword() !== ($originalUser->getPassword() ?? $refreshedUser->getPassword())
+            if (!$originalUser instanceof PasswordAuthenticatedUserInterface || !$refreshedUser instanceof PasswordAuthenticatedUserInterface) {
+                return true;
+            }
+
+            if (null !== ($originalPassword = $originalUser->getPassword())
+                && ($refreshedPassword = $refreshedUser->getPassword()) !== $originalPassword
+                && (8 !== \strlen($originalPassword) || hash('crc32c', $refreshedPassword ?? $originalPassword) !== $originalPassword)
```

Comment:
nitpick: I find the assignment in the condition very hard to read. It's better to have proper assignment.
```suggestion
            $originalPassword = $originalUser->getPassword();
            $refreshedPassword = $refreshedUser->getPassword();
            if (null !== $originalPassword && $refreshedPassword !== $originalPassword
                && (8 !== \strlen($originalPassword) || hash('crc32c', $refreshedPassword ?? $originalPassword) !== $originalPassword)
```

Reply:
updated :)